Fixed issue #20459 by clamping audio stream pitch at a minimum of 0.02f

Antypurus opened the pull request:
Fix for issue https://github.com/godotengine/godot/issues/20459.
From my analysis of this issue it is caused by the audio stream not sending any bytes to the tcp socket when the pitch is very low like 0 or 0.01, however data is expected to be read on the receiver side, when no data is read the generic function assumes that there was a disconnection and issues a disconnect error.
To mitigate this I stop audio mixing when the pitch is lower than the macro `MIN_PITCH_SCALE` in the method `AudioStreamPlayer2D::_mix_audio`. The  `MIN_PITCH_SCALE` has a default value of 0.02 as originally suggested in the bug report issue.

To  test this I created a 2D scene and added the following script to it:
```
extends Node2D

var player
var lower = true

func _ready():
	player = AudioStreamPlayer2D.new()
	self.add_child(player)
	player.stream = load("res://msc.wav")
	player.pitch_scale = 1
	player.play()

func _process(delta):
	if lower:
		player.pitch_scale = player.pitch_scale - 0.01
		if player.pitch_scale<=0.0:
			lower = false
	else:
		if player.pitch_scale < 1.0:
			player.pitch_scale = player.pitch_scale + 0.01
```

As far as my testing goes this handles all values fine including those bellow 0.02 fixing both the game and editor crash.

TiagoJoseMagalhaes commented:
@reduz Hi, do you think this fix is acceptable or should I focus on trying to fix the underlying issue with the audio stream?

mhilbrunner commented:
cc @marcelofg55

marcelofg55 commented on scene/2d/audio_stream_player_2d.cpp:
```diff
@@ -296,6 +297,10 @@ float AudioStreamPlayer2D::get_volume_db() const {
 
 void AudioStreamPlayer2D::set_pitch_scale(float p_pitch_scale) {
+	if (p_pitch_scale < MIN_PITCH_SCALE) {
+		ERR_PRINT("The Pitch Value is bellow the minimum value and as such the audio will be paused");
```
Why not add a return after the ERR_PRINT so that pitch_scale can't be set below the minimum? That way you can remove the check you added on _mix_audio.

TiagoJoseMagalhaes replied:
I though about that and it would definitely be more readable, the reason I didn't do that is so that peoples game doesn't crash if they are changing the pitch the same way i did to test. I thought it would be better. But that is my opinion.
If not crashing is not a need yeah that fix is much better as far as I see it.

TiagoJoseMagalhaes replied:
I added the check so that users could still set those low values for the pitch, and my test script would get stuck on low pitches due to the way I increment the pitch value, however I have tested the way you recommend and it solves the issue as well and if you believe that is the best way, Ill do it, I do believe the way you propose makes the code more readable.

marcelofg55 replied:
I think it's cleaner the way I say, and you can avoid the check at _mix_audio. But maybe somebody else could give another opinion on this too.

TiagoJoseMagalhaes replied:
considering nobody seems to be against your idea I am going to go ahead and implement it

TiagoJoseMagalhaes commented:
@marcelofg55 The travis-ci build for clang seems to be crashing for a math file however I did not touch that file is this something that has been occurring recently? Either ways I have committed the change you proposed.

Noshyaar commented:
Thank you for contributing! The following is just my opinion and not official so don't feel obliged.

1. would be nice to apply the same fix for AudioStreamPlayer and AudioStreamPlayer3D too as they are also affected.
2. I'd propose this change instead:
```
ERR_FAIL_COND(p_pitch_scale < MIN_PITCH_SCALE);
```
I personally find the error message in this specific case too verbose and a bit confusing.
```
The Pitch Value is bellow the minimum value and as such the audio will be paused
```
Since we ignore the new value by returning, the audio won't be paused, right? Also, ERR_FAIL_COND would print the condition that fails so it should be clear that the value is too low. If it's not, I'd use ERR_EXPLAIN on top of it like:
```
ERR_EXPLAIN("Pitch scale below minimum value (0.02)");
ERR_FAIL_COND(p_pitch_scale < MIN_PITCH_SCALE);
```

---

Another fix is to clamp the value below minimum:
```
pitch_scale = p_pitch_scale >= 0.02 ? p_pitch_scale : 0.02;
```
This way we kind of accept the input value rather than outright rejecting it. I used this in a similar issue (value related to audio too low) #16200.

TiagoJoseMagalhaes commented:
@Noshyaar I agree the message is misleading right now as I created it when my approach to solving this was different and have since forgotten to update it. Regarding the suggestion of creating using ERR_FAIL_COND my opinion is that creating a crash for this is unnecessary if we limit the pitch to have value always higher than 0.02. Regarding applying the fix to the other Audio Streams I would be happy to once we land on a fix for this one.
So what I´m thinking is to update the error/warning message to be correct, and not allow values below 0.02 but not crash the game since I think that is unnecessary and might end up being an unneeded annoyance for the user, I feel like the warning should be enough.
What is your opinion on this ?

Noshyaar commented:
` ERR_FAIL_COND` doesn't crash the game. (That's `CRASH_NOW`) It prints message in red if the condition evaluates to true and return immediately. It is used throughout the engine. Example:
https://github.com/godotengine/godot/blob/045f4d99264b9fdf025be7ca243ed5ae0036dfad/scene/2d/audio_stream_player_2d.cpp#L168

I'm kinda biased toward clamping the value as I've used it before for similar situation. It'd be more consistent this way.

TiagoJoseMagalhaes commented:
Ohh my bad I thought it caused a crash, sorry. Then implementing the error as you said seems good.
Clamping seems good to me as well if it is more consistent.
Gonna start working on it then

Noshyaar commented:
Sorry for the confusion. Those are two separated solutions. If you implement clamping, `ERR_FAIL_COND` would have to be removed or it will return prematurely.
I think `WARN_PRINT` or `ERR_PRINT` would work here as the input is altered. I wasn't thinking about it until now that I should have implement some kind of error printing too.

TiagoJoseMagalhaes commented:
Ohh no that is my fault for being  hasty, then I guess the best is to clamp it and add a check that prints a warning about the clamp

Commit:
Fixed game and editor crash related to audio streaming in a 2D scene by clamping the audio pitch with the minimum value of MIN_PITCH_THRESHOLD (currently 0.02f)

Merge remote-tracking branch 'origin/master'


Merge remote-tracking branch 'origin/master'


Update audio_stream_player_2d.cpp
Merge remote-tracking branch 'origin/master'


Update audio_stream_player_2d.cpp
Update audio_stream_player_2d.cpp
update clap behavior

Update audio_stream_player_2d.cpp
Update audio_stream_player_2d.cpp

TiagoJoseMagalhaes commented:
So it seems we've landed on a somewhat good solution, I guess I will move onto applying this fix to the other audio stream classes as suggested by @Noshyaar.

Chaosus commented:
Yeah, just use `CLAMP(p_pitch_scale, 0.01, 32.0);` - because it have this range in the editor. No need to use define for this case

Noshyaar commented:
@TiagoJoseMagalhaes I'm so sorry. I guess the contributors' consensus is shifted from clamping to `ERR_FAIL_COND` (aka what you did before I told you to do something else) and another PR was made. If this is ended up not merged and you want to make another PR, you can fix my clamping in #16200 to also use `ERR_FAIL_COND`. I'll do it myself if you don't want to.

TiagoJoseMagalhaes commented:
@Chaosus during my experimentation, I saw that if you set the pitch to 0.01 through a script it would cause a game and editor crash so that lower limit would not work in my experience. Regarding the use of a define, I am not a fan of the so-called "magic numbers" and that is the reason I use it.
@Noshyaar don't worry for now I will wait for the new PR to unfold and see the consensus before changing anything else in my haste.

Chaosus commented:
Dont pay attention, if you want to make a proper fix - go on, I just didnt detect this crash even with  0.01(probably floating point conversion to 0.0 cause crash)
